ExceptionInInitializerError on failure to read event log

oshi 4.2.0
People getting this exception. No idea how to reproduce.
```
java.lang.ExceptionInInitializerError
    at oshi.SystemInfo.createOperatingSystem(SystemInfo.java:102)
    at oshi.util.Memoizer$1.get(Memoizer.java:84)
    at oshi.SystemInfo.getOperatingSystem(SystemInfo.java:95)
    at launcher.o.j.h(SourceFile:26)
    at launcher.o.j.i(SourceFile:30)
    at launcher.A.a.a(SourceFile:207)
    at launcher.v.e.a(SourceFile:118)
    at dk.xakeps.bootcamp.u.a(SourceFile:241)
    at java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(Unknown Source)
    at java.util.concurrent.ForkJoinTask.doExec(Unknown Source)
    at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(Unknown Source)
    at java.util.concurrent.ForkJoinPool.runWorker(Unknown Source)
    at java.util.concurrent.ForkJoinWorkerThread.run(Unknown Source)
Caused by: com.sun.jna.platform.win32.Win32Exception: Неизвестный интерфейс. (translated: Unknown interface)
    at com.sun.jna.platform.win32.Advapi32Util$EventLogIterator.<init>(Advapi32Util.java:2496)
    at oshi.software.os.windows.WindowsOperatingSystem.querySystemBootTime(WindowsOperatingSystem.java:646)
    at oshi.software.os.windows.WindowsOperatingSystem.<clinit>(WindowsOperatingSystem.java:103)
    ... 13 more
```

--- a/oshi-core/src/main/java/oshi/software/os/windows/WindowsOperatingSystem.java
+++ b/oshi-core/src/main/java/oshi/software/os/windows/WindowsOperatingSystem.java
@@ -650,39 +650,35 @@ public class WindowsOperatingSystem extends AbstractOperatingSystem {
 
     private static long querySystemBootTime() {
         String eventLog = systemLog.get();
-        EventLogIterator iter = null;
         if (eventLog != null) {
             try {
-                iter = new EventLogIterator(null, eventLog, WinNT.EVENTLOG_BACKWARDS_READ);
+                EventLogIterator iter = new EventLogIterator(null, eventLog, WinNT.EVENTLOG_BACKWARDS_READ);
+                // Get the most recent boot event (ID 12) from the Event log. If Windows "Fast
+                // Startup" is enabled we may not see event 12, so also check for most recent ID
+                // 6005 (Event log startup) as a reasonably close backup.
+                long event6005Time = 0L;
+                while (iter.hasNext()) {
+                    EventLogRecord record = iter.next();
+                    if (record.getStatusCode() == 12) {
+                        // Event 12 is system boot. We want this value unless we find two 6005 events
+                        // first (may occur with Fast Boot)
+                        return record.getRecord().TimeGenerated.longValue();
+                    } else if (record.getStatusCode() == 6005) {
+                        // If we already found one, this means we've found a second one without finding
+                        // an event 12. Return the latest one.
+                        if (event6005Time > 0) {
+                            return event6005Time;
+                        }
+                        // First 6005; tentatively assign
+                        event6005Time = record.getRecord().TimeGenerated.longValue();
+                    }
+                }
+                // Only one 6005 found, return
+                if (event6005Time > 0) {
+                    return event6005Time;
+                }
             } catch (Win32Exception e) {
                 LOG.warn("Can't open event log \"{}\".", eventLog);
-            }
-        }
-        // If no event log, subtract up time
-        if (eventLog != null && iter != null) {
-            // Get the most recent boot event (ID 12) from the Event log. If Windows "Fast
-            // Startup" is enabled we may not see event 12, so also check for most recent ID
-            // 6005 (Event log startup) as a reasonably close backup.
-            long event6005Time = 0L;
-            while (iter.hasNext()) {
-                EventLogRecord record = iter.next();
-                if (record.getStatusCode() == 12) {
-                    // Event 12 is system boot. We want this value unless we find two 6005 events
-                    // first (may occur with Fast Boot)
-                    return record.getRecord().TimeGenerated.longValue();
-                } else if (record.getStatusCode() == 6005) {
-                    // If we already found one, this means we've found a second one without finding
-                    // an event 12. Return the latest one.
-                    if (event6005Time > 0) {
-                        return event6005Time;
-                    }
-                    // First 6005; tentatively assign
-                    event6005Time = record.getRecord().TimeGenerated.longValue();
-                }
-            }
-            // Only one 6005 found, return
-            if (event6005Time > 0) {
-                return event6005Time;
             }
         }
         // If we get this far, event log reading has failed, either from no log or no
@@ -754,7 +750,7 @@ public class WindowsOperatingSystem extends AbstractOperatingSystem {
     }
 
     private static String querySystemLog() {
-        String systemLog = GlobalConfig.get("oshi.os.windows.systemLog", "System");
+        String systemLog = GlobalConfig.get("oshi.os.windows.eventlog", "System");
         if (systemLog.isEmpty()) {
             // Use faster boot time approximation
             return null;
@@ -762,7 +758,7 @@ public class WindowsOperatingSystem extends AbstractOperatingSystem {
         // Check whether it works
         HANDLE h = Advapi32.INSTANCE.OpenEventLog(null, systemLog);
         if (h == null) {
-            LOG.warn("Unable to locate configured system Event log \"{}\". Calculating boot time from uptime.",
+            LOG.warn("Unable to open configured system Event log \"{}\". Calculating boot time from uptime.",
                     systemLog);
             return null;
         }

--- a/oshi-core/src/main/java/oshi/software/os/windows/WindowsOperatingSystem.java
+++ b/oshi-core/src/main/java/oshi/software/os/windows/WindowsOperatingSystem.java
@@ -26,6 +26,7 @@ package oshi.software.os.windows;
 import static oshi.software.os.OSService.State.OTHER;
 import static oshi.software.os.OSService.State.RUNNING;
 import static oshi.software.os.OSService.State.STOPPED;
+import static oshi.util.Memoizer.memoize;
 
 import java.io.File;
 import java.util.ArrayList;
@@ -36,6 +37,8 @@ import java.util.HashSet;
 import java.util.List;
 import java.util.Map;
 import java.util.Set;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Supplier;
 
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
@@ -86,6 +89,7 @@ import oshi.software.os.OSProcess;
 import oshi.software.os.OSService;
 import oshi.software.os.OSService.State;
 import oshi.software.os.OperatingSystem;
+import oshi.util.GlobalConfig;
 import oshi.util.ParseUtil;
 import oshi.util.platform.windows.PerfCounterQuery;
 import oshi.util.platform.windows.PerfCounterWildcardQuery;
@@ -99,6 +103,12 @@ public class WindowsOperatingSystem extends AbstractOperatingSystem {
 
     private static final boolean IS_VISTA_OR_GREATER = VersionHelpers.IsWindowsVistaOrGreater();
     private static final boolean IS_WINDOWS7_OR_GREATER = VersionHelpers.IsWindows7OrGreater();
+
+    /**
+     * Windows event log name
+     */
+    private static Supplier<String> systemLog = memoize(WindowsOperatingSystem::querySystemLog,
+            TimeUnit.HOURS.toNanos(1));
 
     private static final long BOOTTIME = querySystemBootTime();
 
@@ -639,30 +649,45 @@ public class WindowsOperatingSystem extends AbstractOperatingSystem {
     }
 
     private static long querySystemBootTime() {
-        // Get the most recent boot event (ID 12) from the Event log. If Windows "Fast
-        // Startup" is enabled we may not see event 12, so also check for most recent ID
-        // 6005 (Event log startup) as a reasonably close backup.
-        long event6005Time = 0L;
-        EventLogIterator iter = new EventLogIterator(null, "System", WinNT.EVENTLOG_BACKWARDS_READ);
-        while (iter.hasNext()) {
-            EventLogRecord record = iter.next();
-            if (record.getStatusCode() == 12) {
-                // Event 12 is system boot. We want this value unless we find two 6005 events
-                // first (may occur with Fast Boot)
-                return record.getRecord().TimeGenerated.longValue();
-            } else if (record.getStatusCode() == 6005) {
-                // If we already found one, this means we've found a second one without finding
-                // an event 12. Return the latest one.
-                if (event6005Time > 0) {
-                    return event6005Time;
-                }
-                // First 6005; tentatively assign
-                event6005Time = record.getRecord().TimeGenerated.longValue();
+        String eventLog = systemLog.get();
+        EventLogIterator iter = null;
+        if (eventLog != null) {
+            try {
+                iter = new EventLogIterator(null, eventLog, WinNT.EVENTLOG_BACKWARDS_READ);
+            } catch (Win32Exception e) {
+                LOG.warn("Can't open event log \"{}\".", eventLog);
             }
         }
-        // If we get this far, event log reading has failed. Subtract up time from
-        // current time as a reasonable proxy.
-        return event6005Time > 0 ? event6005Time : System.currentTimeMillis() / 1000L - querySystemUptime();
+        // If no event log, subtract up time
+        if (eventLog != null && iter != null) {
+            // Get the most recent boot event (ID 12) from the Event log. If Windows "Fast
+            // Startup" is enabled we may not see event 12, so also check for most recent ID
+            // 6005 (Event log startup) as a reasonably close backup.
+            long event6005Time = 0L;
+            while (iter.hasNext()) {
+                EventLogRecord record = iter.next();
+                if (record.getStatusCode() == 12) {
+                    // Event 12 is system boot. We want this value unless we find two 6005 events
+                    // first (may occur with Fast Boot)
+                    return record.getRecord().TimeGenerated.longValue();
+                } else if (record.getStatusCode() == 6005) {
+                    // If we already found one, this means we've found a second one without finding
+                    // an event 12. Return the latest one.
+                    if (event6005Time > 0) {
+                        return event6005Time;
+                    }
+                    // First 6005; tentatively assign
+                    event6005Time = record.getRecord().TimeGenerated.longValue();
+                }
+            }
+            // Only one 6005 found, return
+            if (event6005Time > 0) {
+                return event6005Time;
+            }
+        }
+        // If we get this far, event log reading has failed, either from no log or no
+        // startup times. Subtract up time from current time as a reasonable proxy.
+        return System.currentTimeMillis() / 1000L - querySystemUptime();
     }
 
     @Override
@@ -726,6 +751,22 @@ public class WindowsOperatingSystem extends AbstractOperatingSystem {
             LOG.error("Win32Exception: {}", ex.getMessage());
             return new OSService[0];
         }
+    }
+
+    private static String querySystemLog() {
+        String systemLog = GlobalConfig.get("oshi.os.windows.systemLog", "System");
+        if (systemLog.isEmpty()) {
+            // Use faster boot time approximation
+            return null;
+        }
+        // Check whether it works
+        HANDLE h = Advapi32.INSTANCE.OpenEventLog(null, systemLog);
+        if (h == null) {
+            LOG.warn("Unable to locate configured system Event log \"{}\". Calculating boot time from uptime.",
+                    systemLog);
+            return null;
+        }
+        return systemLog;
     }
 
     enum OSVersionProperty {

--- a/oshi-core/src/main/resources/oshi.properties
+++ b/oshi-core/src/main/resources/oshi.properties
@@ -29,6 +29,21 @@ oshi.util.proc.path=/proc
 # Default is -1, no timeout
 oshi.util.wmi.timeout=-1
 
+# The name of the System event log containing bootup event IDs 12 and 6005.
+# This is used for a one-time calculation of system boot time that should be
+# consistent across process runs regardless of sleep/hibernate cycles, at 
+# the small cost of ~250ms latency reading upon WindowsOperatingSystem
+# initialization. 
+#
+# If this variable doesn't match or is left as the empty string, boot time
+# will be calculated by subtracting up time from current time. This may vary
+# by up to a millisecond between program executions and does not properly 
+# account for sleep/hibernate cycles, but is extremely low latency and may
+# be preferred if only approximate boot time is desired.
+#  
+# Default is System
+oshi.os.windows.systemLog=System
+
 # Memoizer default expiration in milliseconds (return values will be cached this long)
 # Must be positive (negative value will never refresh)
 # Should be less than 1 second

--- a/oshi-core/src/main/resources/oshi.properties
+++ b/oshi-core/src/main/resources/oshi.properties
@@ -29,20 +29,22 @@ oshi.util.proc.path=/proc
 # Default is -1, no timeout
 oshi.util.wmi.timeout=-1
 
-# The name of the System event log containing bootup event IDs 12 and 6005.
+# The name of the System event log containing bootup event IDs 12 and 6005. If a 
+# non-empty invalid log name is specified, the name "Application" will be used.
+#
 # This is used for a one-time calculation of system boot time that should be
 # consistent across process runs regardless of sleep/hibernate cycles, at 
 # the small cost of ~250ms latency reading upon WindowsOperatingSystem
 # initialization. 
 #
-# If this variable doesn't match or is left as the empty string, boot time
-# will be calculated by subtracting up time from current time. This may vary
-# by up to a millisecond between program executions and does not properly 
-# account for sleep/hibernate cycles, but is extremely low latency and may
-# be preferred if only approximate boot time is desired.
-#  
+# If the specified log is the empty string, or doesn't contain a bootup event,
+# boot time will be calculated by subtracting up time from current time. This 
+# may vary by up to a millisecond between program executions and does not 
+# properly account for sleep/hibernate cycles, but is fast and may be preferred 
+# if only approximate boot time is desired.
+#
 # Default is System
-oshi.os.windows.systemLog=System
+oshi.os.windows.eventlog=System
 
 # Memoizer default expiration in milliseconds (return values will be cached this long)
 # Must be positive (negative value will never refresh)

